### DISCUSSION TITLE
Imgur v2.0.0 (unofficial) released!

### DISCUSSION THREAD

**Discussion by KenEucker** [KenEucker/imgur#105] [Announcements]
<img width="1188" alt="imgurV2" src="https://user-images.githubusercontent.com/1720267/147290319-ec92a677-d834-4015-9590-3dc3419522de.png">

**It's here!**

Imgur v2.0.0 is now ready for you to use in the wild! This is a preliminary release to get the next version of the Imgur API out there. The biggest notes for this release are:

* Entirely new code, written 100% in TypeScript
* This package is now compiled into an isomorphic library that works both in NodeJS and in the browser
* The maintenance has moved from: https://github.com/kaimallea/node-imgur to this repository: https://github.com/KenEucker/imgur
* The implicit use of the `fs` node module has been removed

Many thanks to @kaimallea for starting this work (vNext) and for ushering it along to a new future. Many thanks to everyone else who has contributed to this project over the last 9+ years of its life. Many thanks to those 2,800+ weekly downloads that use the Imgur API out in the world of Open-Source development. 

Please let me know if you want to be involved in the further success of this project by leaving a comment below.

**Comment by efraimbart** [comment]
A couple notes on this update:

1. [FormData.getHeaders](https://github.com/KenEucker/imgur/blob/main/src/image/upload.ts#L26) is not isomorphic and fails in the browser
2. The imgur `/upload` endpoint is blocked by cors in the browser, perhaps we need to use the undocumented `/image` endpoint?
3. There is some nullish coalescing and optional chaining in the compiled imgur.js file. That doesn't work everywhere.
4. When attempting to upload from NodeJS, if attempting via the `stream` key it fails, apparently because it appends the [whole payload](https://github.com/KenEucker/imgur/blob/main/src/common/utils.ts#L56) to the form instead of just the stream. If attempting via `video` key it fails as well, perhaps because the `filename` is not passed in.

On the documentation side, I'm not sure I understand some things around [stream](https://github.com/KenEucker/imgur/blob/main/README.md?plain=1#L91):

1. The documentation seems to replace `file` with `stream`, and apparently the app even attempts to send it to the imgur api as such.
2. There is no equivalent `stream` option for videos.

**Reply by efraimbart** [reply]
In short, I've been unable to use this to actually upload anything to imgur.

**Reply by KenEucker** [reply]
@efraimbart I will look into this later today. I have been using this to make image uploads to Imgur using the code in the example. Would you be open to working on an example branch of this repo together?

**Reply by efraimbart** [reply]
Sure! 

The working example is with the path to the image, correct? I think it's the other options that don't work.

**Reply by KenEucker** [reply]
@efraimbart you're welcome to fork this repo and start that branch if you like.

I am curious if you are able to upload using v1 of `imgur`? If so, that would be a good place to start.

**Reply by efraimbart** [reply]
I don't know if I'll get around to starting the example branch myself in the near future, but if it's started I can definitely contribute.

I have not yet tried via v1, but I did manage to get uploads working by modifying the `imgur.js` and `imgur.node.js` files, so I can confirm that it would work.

**Reply by efraimbart** [reply]
Example of a fix for 4. above https://github.com/KenEucker/imgur/pull/110.

**Reply by KenEucker** [reply]
Thanks for the PR, I will pull it into develop and we can start working on the example repo later. I can also create the necessary example branch.

**Comment by KenEucker** [comment]
@efraimbart 

1. I believe that the `form-data` dependency is polyfilled by webpack during compilation, but I could be misunderstanding. I know that form-data is not isomorphic and had originally intended on a path to remove it, but found that it worked when compiling using webpack and I haven't encountered any issues in the browser.
2. I believe that you can overcome the CORS issue by adding referer=no-referrer to the page meta. We should test for this in the example branch to be created.
3. This is due to the compile target being `es2020`. In order to support older versions of javascript, we would target `es2018` or earlier. I am unsure about the right decision here, for the moment.
4. It seems that your PR addresses this, correct?

**Reply by efraimbart** [reply]
1. It doesn't work for me, I guess we can use the example repo to track down this discrepancy.
2. I can definitely try that.
3. `vue-loader` throws a fit because of this. Just something to keep in mind when making your decision
4. My PR is very much a sample PR in that it makes one change to fix one thing without taking into account the surroundings. For example, it only fixes videos - not images, it doesn't address the confusion I have around `stream` - rather circumvents it, and it doesn't update the documentation.

**Reply by KenEucker** [reply]
@efraimbart I have added an example folder to the `develop` branch that has a working example for uploads of both image and video using streams. Here is the code:
https://github.com/KenEucker/imgur/blob/develop/example/index.js

Please run this example on your machine and let me know if you are able to get it working. You will need to create a .env file with the following fields:
```
CLIENT_ID=
ACCESS_TOKEN=
ALBUM=
```

I believe that the access token is required to make uploads to a specific user folder (and also permissions to do so by that user). Without these fields it uploads it publicly not attached to any account, so it should be testable with just the clientId.

**Reply by KenEucker** [reply]
@efraimbart I just realized the issue while circling back to the README. In the README I had improperly instructed to use a field 'stream' which you provided that fix for, however, the correct field is still `image` with the `type` set to 'stream'. My apologies for that and for any lost time that working around this has caused you!

I can publish a new version from the current `develop` branch after we confirm this is resolved on your end.

EDIT: I added browser code to the example. I can now reproduce your issue with formData.getHeader using the compiled library. I added [example/index.html](https://github.com/KenEucker/imgur/blob/develop/example/index.html) which needs to be served in order to work ( I use [vscode live server](https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer) ).

**Reply by KenEucker** [reply]
@efraimbart my recent updates to the `develop` branch resolve the getHeaders issue. ~I am now trying to sort out how to properly create a stream in the browser that can be used in the upload.~ I have sorted out how to upload streams in the browser and have a working example for both images and videos. However, I appear to be getting a CORS error on uploading the video (the `/upload` endpoint has CORS issues where the `/image` endpoint does not). Despite the error thrown from axios, I still see the video gets uploaded.

If you have a chance, take a look at [example/index.html](https://github.com/KenEucker/imgur/blob/develop/example/index.html) and see if you can get it to work?

**Reply by KenEucker** [reply]
I am going to publish 2.0.1 right now, as I am also trying to use this package for uploads in another project and it appears to be working.

**Reply by efraimbart** [reply]
Sorry for going awol on you, I should have a chance to work on this tomorrow
